sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c

```diff
@@ -120,13 +121,15 @@ static int _check_fragments(gnrc_sixlowpan_frag_rb_base_t *entry,
     return RBUF_ADD_SUCCESS;
 }
 
-void gnrc_sixlowpan_frag_rb_add(gnrc_netif_hdr_t *netif_hdr,
-                                gnrc_pktsnip_t *pkt, size_t offset,
-                                unsigned page)
+gnrc_sixlowpan_frag_rb_t *gnrc_sixlowpan_frag_rb_add(gnrc_netif_hdr_t *netif_hdr,
+                                                     gnrc_pktsnip_t *pkt,
+                                                     size_t offset, unsigned page)
 {
-    if (_rbuf_add(netif_hdr, pkt, offset, page) == RBUF_ADD_REPEAT) {
-        _rbuf_add(netif_hdr, pkt, offset, page);
+    int res;
+    if ((res = _rbuf_add(netif_hdr, pkt, offset, page)) == RBUF_ADD_REPEAT) {
+        res = _rbuf_add(netif_hdr, pkt, offset, page);
```

Comment:
I  suppose `RBUF_ADD_REPEAT` can only be returned once?

Reply:
TL;DR Yes. This call was originally a recursion which was unrolled in 7452f48c2608d73d715aeeba23c8b5f2d8c49c0a. This line reflects basically the marked part of the following paragraph in [RFC 4944, section 5.3](https://tools.ietf.org/html/rfc4944#section-5.3):

>    If a link fragment that overlaps another fragment is received, as
>    identified above, and differs in either the size or datagram_offset
>    of the overlapped fragment, the fragment(s) already accumulated in
>    the reassembly buffer SHALL be discarded.  **A fresh reassembly may be
>    commenced with the most recently received link fragment.**  Fragment
>    overlap is determined by the combination of datagram_offset from the
>    encapsulation header and "Frame Length" from the 802.15.4 Physical
>    Layer Protocol Data Unit (PPDU) packet header.

So `_rbuf_add()` only returns `RBUF_ADD_REPEAT` if the new fragment overlaps with an existing one. In that case however, the reassembly buffer entry gets discarded (with all its collected fragments so far), so re-calling `_rbuf_add()` creates a new entry with the previously overlapping fragment as its only entry.

Reply:
Makes sense. I think it doesn't hurt to add a short comment about that so nobody will try to 'fix' this in the future.

Reply:
Didn't really fit with the existing commits, so I just added a separate commit to add the comment.